staging/src/k8s.io/mount-utils/mount_windows.go

```diff
@@ -288,8 +288,12 @@ func (mounter *SafeFormatAndMount) formatAndMountSensitive(source string, target
 	}
 
 	// format disk if it is unformatted(raw)
+	formatOptionsUnwrapped := ""
+	if len(formatOptions) > 0 {
+		formatOptionsUnwrapped = " " + strings.Join(formatOptions, " ")
+	}
 	cmd := fmt.Sprintf("Get-Disk -Number %s | Where partitionstyle -eq 'raw' | Initialize-Disk -PartitionStyle GPT -PassThru"+
-		" | New-Partition -UseMaximumSize | Format-Volume -FileSystem %s -Confirm:$false", source, fstype)
+		" | New-Partition -UseMaximumSize | Format-Volume -FileSystem %s -Confirm:$false%s", source, fstype, formatOptionsUnwrapped)
```

Comment:
I think this will work for an intree driver but not for a CSI driver that uses CSI Proxy to do the format & mount

Reply:
Is there a way how to pass such parameters through CSI proxy?
We can document that this will work only when a CSI driver is running as a "privileged container", without the proxy. Will it work then?

Reply:
Yes there's a way to pass parameters to CSI Proxy, CSI Drivers use the API exposed by CSI Proxy, currently it formats only with the volumeId https://github.com/kubernetes-csi/csi-proxy/blob/master/client/api/volume/v1/api.proto#L22 in the proto definition but as we're in the process to move to HostProcess containers this would need to go here instead: https://github.com/kubernetes-csi/csi-proxy/blob/4624937be58bfd46f5ebe6d8f213f58414275e45/pkg/volume/volume.go#L140

After reading the PR in more detail it's doing the following in Windows:

```
	formatOptionsUnwrapped := ""
	if len(formatOptions) > 0 {
		formatOptionsUnwrapped = " " + strings.Join(formatOptions, " ")
	}
```

Is the expectation of the format additional options to work in Windows? I don't think they will work even in intree code, I'd suggest removing this part for Windows and only it in Linux.

To enable something similar in Windows there needs to be a mapping of a `mkfs` format option to a `Format-Volume` format option.

Reply:
Good questions!

> Is the expectation of the format additional options to work in Windows?

Yes, I would say so. The code as it's suggested in this PR moves this responsibility to the caller (= CSI driver / in-tree volume plugin / user). It must provide the right options for the right OS and the right fsType. mount-utils should be IMO small and stupid in this regard.

@ConnorJC3, so... if user wants a ntfs volume that will be usable both on Windows and Linux nodes, they will need to provide both `mkfs` and `Format-Volume` options, because they don't know what node will format the volume (i.e. if a Windows or Linux Pod gets the PV first). How will the `StorageClass.Parameters` will look like? Will be the driver really smart enough to choose the right options? It is solvable, but it's more complex that I initially thought.

> I don't think they will work even in intree code, I'd suggest removing this part for Windows and only use it in Linux.

I am not a Windows person, @ConnorJC3, does it sound right? Can you check it, update the PR as suggested and update csi-proxy?

Reply:
So as far as I know, there are 2 options here:

1. Use a privileged container and directly manipulate the volumes. If a driver is doing this (and uses `mount-utils`), it should work.
2. Use the `csi-proxy` daemon. Basically everybody does this right now, because up until recently privileged containers on Windows did not exist. We (as in the aws-ebs-csi-driver) plan to eventually move to using them, but have to wait for now until all our supported platforms can use them.

In case number 2, changes would have to be made to the CSI Proxy in order to support this for drivers. However, I'm not sure if it's worth doing so given that the CSI Proxy should be on its way out anyways. We'll probably hold off unless it's specifically requested.

Either way, as `mount-utils` will become the de-facto place for this implementation as privileged Windows containers are phased in, we should keep it here.

Reply:
> @ConnorJC3, so... if user wants a ntfs volume that will be usable both on Windows and Linux nodes, they will need to provide both mkfs and Format-Volume options, because they don't know what node will format the volume (i.e. if a Windows or Linux Pod gets the PV first). How will the StorageClass.Parameters will look like? Will be the driver really smart enough to choose the right options? It is solvable, but it's more complex that I initially thought.

I'm not sure if this is a huge problem, because this should _already_ be true with mount options, and nobody's complained about those as far as I know. Also, this is more of a discussion that makes sense on the CSI Driver side, `mount-utils` just provides the implementation for formatting and mounting volumes.

1. If a CSI Driver is creating the format options (say, for example, the driver had an option for block size and mapped that to the relevant format option), then it could just do it correctly for each platform.
2. If the user is directly supplying the format options, then the driver would probably need to take a set of options for both linux and windows, and use the correct one depending on which platform the formatting ran on.

Reply:
> As mount-utils will become the de-facto place for this implementation as privileged Windows containers are phased in, we should keep it here.

That's my understanding too, but again, I know very little about Windows. @mauriciopoppe, is the current PR correct then?

Reply:
> As mount-utils will become the de-facto place for this implementation as privileged Windows containers are phased in, we should keep it here

This is still up for discussion, I think it's worth talking about it with the community.

> In case number 2, changes would have to be made to the CSI Proxy in order to support this for drivers. However, I'm not sure if it's worth doing so given that the CSI Proxy should be on its way out anyways.

With HostProcess containers CSI Proxy is going to become a library and will be used in the CSI Drivers like mount-utils , this is tracked in this KEP https://github.com/kubernetes/enhancements/pull/3641, @ddebroy and I had a discussion about it in the CSI Windows meeting, quoting some part of it: ["reasons to keep CSI Proxy independent, CSI Proxy is more than mount e.g. ops for SMB, iSCSI, broader in the number of problems it solves, it's like storage utils than mount-utils."](https://docs.google.com/document/d/1_WtQUnnlXE_M-YzYVrPZbjEgSxMvY9EYFwl61MC2i5I/edit#bookmark=id.m6yzw8n3okss)

I think I should take back what I said in https://github.com/kubernetes/kubernetes/pull/112877#discussion_r1010779048 about it not working in Windows, I think it'll work and enable sending additional flags which is useful and it's all in the control of the caller which should have Windows/Linux variations in their mount code.

/remove-hold